### PR TITLE
Filter queried Features for map way name

### LINKED ISSUE #1016: Add check to reduce way name query sensitivity / update frequency
We are currently querying the map data for the current way name that is being traveled along.  The querying is very precise and this sometimes causes the way name to update too frequently.  

We should add a check (similar to [iOS](https://github.com/mapbox/mapbox-navigation-ios/blob/master/MapboxNavigation/RouteMapViewController.swift#L782-L807)) to look forward and confirm that this change in way name is actually warranted.  

ezgif com-video-to-gif

### PR DESCRIPTION
**Pull request by danesfeder**
Closes #1016 

Adds check to stop updating of map way name when crossing through streets or other roads unnecessarily.

### CODE REVIEW

File: libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/map/WaynameFeatureFilter.java [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+  private double calculateDistance(Point pointAheadUser, Point pointAheadFeature) {
+    if (pointAheadUser == null || pointAheadFeature == null) {
+      return Double.POSITIVE_INFINITY;
+    }
+    return TurfMeasurement.distance(pointAheadUser, pointAheadFeature);
+  }
+
+  @Nullable
+  Point findPointFromCurrentPoint(Point currentPoint, double metersFromCurrentPoint, LineString lineString) {
```
Comment (danesfeder) [inline libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/map/WaynameFeatureFilter.java]: @1ec5 @vincethecoder I believe I have most of the logic here, minus this method.  iOS is using a chain of extension functions to accomplish what I'm trying to do here in this method.  

This method is called three times starting https://github.com/mapbox/mapbox-navigation-android/pull/1156/files#diff-1e717bd96de82a3e257f8b4d8e92e0a2R69.  Do you see any issues with this logic?  

When it's called it looks like it returns the same values for all three scenarios (feature ahead, feature behind, user ahead):
```
D/WaynameFeatureFilter: NAV_DEBUG pointAheadFeature: Point{type=Point, bbox=null, coordinates=[-77.02962967632158, 38.904698288477306]}
D/WaynameFeatureFilter: NAV_DEBUG pointBehindFeature: Point{type=Point, bbox=null, coordinates=[-77.02962967632158, 38.904698288477306]}
D/WaynameFeatureFilter: NAV_DEBUG pointAheadUser: Point{type=Point, bbox=null, coordinates=[-77.0296298615751, 38.90469829037757]}
```

**Comment by danesfeder** [comment]
Originally opened as https://github.com/mapbox/mapbox-navigation-android/issues/1227:

@Guardiola31337 Run into the following 💥 while testing https://github.com/mapbox/mapbox-navigation-android/pull/1156

```
08-20 13:14:02.849 15537-15537/com.mapbox.services.android.navigation.testapp E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.mapbox.services.android.navigation.testapp, PID: 15537
    java.lang.NullPointerException: Attempt to invoke virtual method 'boolean com.mapbox.geojson.Feature.hasNonNullValueForProperty(java.lang.String)' on a null object reference
        at com.mapbox.services.android.navigation.ui.v5.map.MapWayname.updateWaynameLayerWithNameProperty(MapWayname.java:140)
        at com.mapbox.services.android.navigation.ui.v5.map.MapWayname.updateLayerWithRoadLabelFeatures(MapWayname.java:108)
        at com.mapbox.services.android.navigation.ui.v5.map.MapWayname.updateWaynameWithPoint(MapWayname.java:59)
        at com.mapbox.services.android.navigation.ui.v5.map.NavigationMapboxMap.updateMapWaynameWithLocation(NavigationMapboxMap.java:268)
        at com.mapbox.services.android.navigation.ui.v5.map.NavigationMapboxMap.updateLocation(NavigationMapboxMap.java:89)
        at com.mapbox.services.android.navigation.ui.v5.NavigationView.updateNavigationMap(NavigationView.java:331)
        at com.mapbox.services.android.navigation.ui.v5.NavigationPresenter.onNavigationLocationUpdate(NavigationPresenter.java:66)
        at com.mapbox.services.android.navigation.ui.v5.NavigationViewSubscriber$3.onChanged(NavigationViewSubscriber.java:43)
        at com.mapbox.services.android.navigation.ui.v5.NavigationViewSubscriber$3.onChanged(NavigationViewSubscriber.java:39)
        at android.arch.lifecycle.LiveData.considerNotify(LiveData.java:109)
        at android.arch.lifecycle.LiveData.dispatchingValue(LiveData.java:126)
        at android.arch.lifecycle.LiveData.setValue(LiveData.java:282)
        at android.arch.lifecycle.MutableLiveData.setValue(MutableLiveData.java:33)
        at com.mapbox.services.android.navigation.ui.v5.NavigationViewModel$1.onProgressChange(NavigationViewModel.java:267)
        at com.mapbox.services.android.navigation.v5.navigation.NavigationEventDispatcher.onProgressChange(NavigationEventDispatcher.java:144)
        at com.mapbox.services.android.navigation.v5.navigation.RouteProcessorThreadListener.onNewRouteProgress(RouteProcessorThreadListener.java:34)
        at com.mapbox.services.android.navigation.v5.navigation.RouteProcessorHandlerCallback$1.run(RouteProcessorHandlerCallback.java:97)
        at android.os.Handler.handleCallback(Handler.java:739)
        at android.os.Handler.dispatchMessage(Handler.java:95)
        at android.os.Looper.loop(Looper.java:148)
        at android.app.ActivityThread.main(ActivityThread.java:5417)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
```

With this new code, the filter has the ability to return `null` and we need to account for this [here](https://github.com/mapbox/mapbox-navigation-android/pull/1156/files#diff-1a05354399a29ce52986ed6514beccd5R107) before the `Feature` is passed to `updateWaynameLayerWithNameProperty()`.  Good catch!

**Comment by Guardiola31337** [comment]
Following are my findings after debugging / testing the latest "working" changes:
- Performance 📉 The puck is jumpy when features are filtered 👀 

query_sensitivity_filter_jumpy

This is even worse when a bunch of road features need to be filtered. Noting that we're executing all this logic on each location update (every second). Maybe we should do it on a different thread? cc @danesfeder 
- Saw some incorrect waynames when the camera is zoomed out. I guess this is because zoom levels affect the query rendered features returned. https://github.com/mapbox/mapbox-navigation-android/blob/0854e8e0c17113eef980fafe1dbc8a290f5dd35f/libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/map/WaynameFeatureFinder.java#L18-L20
@1ec5 could you clarify?
- Sometimes when there's only one feature to filter without changing roads at any moment https://github.com/mapbox/mapbox-navigation-android/blob/7cd6d9f846163d8fcf041004781fd4c1ba066744/libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/map/WaynameFeatureFilter.java#L52-L56 the wayname doesn't match the immediate previous one.
- Found situations in which the road features obtained are the "same" 👀 
```
NAV_DEBUG Queried names: [Paseo de Santa María de la Cabeza, Paseo de Santa María de la Cabeza]
```
I believe this is because that street has multiple roads (2 in this concrete case) but I'm wondering if we could filter the repeated ones out, avoiding unnecessary filtering steps and improving the overall performance. I've also seen multiple streets repeated 👀
```
NAV_DEBUG Queried names: [Paseo de Santa María de la Cabeza, Paseo de Santa María de la Cabeza, Autovía de Toledo, Autovía de Toledo]
```
- What happens in the remote case in which `minDistanceToFeature` is equal to `smallestUserDistanceToFeature`? Could this be a problem i.e. return a wrong filtered feature?
https://github.com/mapbox/mapbox-navigation-android/blob/7cd6d9f846163d8fcf041004781fd4c1ba066744/libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/map/WaynameFeatureFilter.java#L141-L145

Please let me know if something stated above is incorrect or missing 🙏

---

TODOs
- [ ] Remove debugging logs
- [ ] Fix tests
- [ ] Consider / revisit https://github.com/mapbox/mapbox-navigation-android/blob/7cd6d9f846163d8fcf041004781fd4c1ba066744/libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/map/WaynameFeatureFilter.java#L44 https://github.com/mapbox/mapbox-navigation-android/blob/7cd6d9f846163d8fcf041004781fd4c1ba066744/libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/map/MapWayname.java#L104 See https://github.com/mapbox/mapbox-navigation-android/pull/1156#issuecomment-414659621

Feel free to add here or as a comment anything I might have missed.

**Comment by Guardiola31337** [comment]
Follow up questions / comments
- All the "reversing" calculations would be a lot easier if [`TurfMeasurement#along`](https://github.com/mapbox/mapbox-java/blob/master/services-turf/src/main/java/com/mapbox/turf/TurfMeasurement.java#L232-L264) would accept negative distances. Wondering why this is not currently an option. Something technically not feasible or not implemented yet? 🤔 cc @osana 
- Could the Directions API keep track of the wayname somehow so we can get it from the leg step as simple as `LegStep#currentWayName`? Thinking on options to take out this expensive filtering logic out from the client cc @mapbox/directions @danesfeder

**Comment by 1ec5** [comment]
> To clarify this further, let me give an example. Let's say that we're in a route from A to B; we're moving towards B but when we query the feature the first point of the line string (first position of the list) is a point close to B and we were using always the last point of the list as the point "moving ahead" which was a point close to A (opposite direction / behind) messing all the calculations up. In this example the street numbers start in B which means (B) 1, 2, 3, …, (A) increasing towards A; because we don’t know that beforehand, we need to calculate it / be aware of; 👇 is the block of code managing this logic

You’re right; in short, the route could travel in either direction along a two-way street, but the vector tile would contain the same data either way. This is why the code calculates both the direction “ahead” and the direction “behind”. These variables are poorly named: nothing behind the user actually matters to us; “behind” in this case means “[backwards](https://wiki.openstreetmap.org/wiki/Forward_%26_backward,_left_%26_right) along the feature in the vector tile”, which could be forward along the route.

**Comment by 1ec5** [comment]
> Saw some incorrect waynames when the camera is zoomed out. I guess this is because zoom levels affect the query rendered features returned.

We have no truly reliable way to determine the name of the current road. The most obvious solution would be to get the name from the current step in the route model, but the Directions API often consolidates steps without regard to the exact road names. We resort to querying Mapbox Streets vector tiles, since the Streets source only unions together features with the same exact name. Unfortunately, feature querying can be imprecise, since the tiles don’t retain the road topology, only the geometries. We mitigate the issue to some extent with heuristics based on candidate features’ bearings, but we’re unable to make the feature fully reliable until the Directions API is able to reliably tell us when a road name changes along the route.

/ref mapbox/mapbox-navigation-ios#590

**Comment by Guardiola31337** [comment]
@osana 
> [`TurfMeasurement#along`](https://github.com/mapbox/mapbox-java/blob/master/services-turf/src/main/java/com/mapbox/turf/TurfMeasurement.java#L232-L264)  takes different unit parameters. Would passing degrees or radians help you?

I was referring to `@FloatRange(from = 0) double distance` which states that the distance needs to be a float starting from 0 (no negative numbers allowed). That is because how `along` is implemented 
```java
* @return a {@link Point} which is on the linestring provided and at the distance from the origin
*   of that line to the end of the distance
```
and I was wondering if we have ever considered adding support for negative distances i.e. if the distance is negative return a point which is on the linestring provided and at the distance from the **end** of that line to the **origin** of the distance. That'd remove the necessity of "reversing" the lines in which a "behind" point in the line is needed + would make all the calculations a lot simpler. Is there something technically not feasible here or we haven't thought about it until now?

**Comment by danesfeder** [comment]
@Guardiola31337 I updated this to use an `AsyncTask` to run the filter and it seems to have worked pretty well - do you mind testing / taking another look here?  Thanks!

**Comment by Guardiola31337** [comment]
@danesfeder 
> I updated this to use an `AsyncTask` to run the filter and it seems to have worked pretty well - do you mind testing / taking another look here?  Thanks! 

Took another pass here and it's looking good 🎉 

One quick win that we can add to make it 💯 before merging is as commented above https://github.com/mapbox/mapbox-navigation-android/pull/1156#issuecomment-415385727

> - Found situations in which the road features obtained are the "same" 👀 
>```
>NAV_DEBUG Queried names: [Paseo de Santa María de la Cabeza, Paseo de Santa María de la Cabeza]
>```
>I believe this is because that street has multiple roads (2 in this concrete case) but I'm wondering if we could filter the repeated ones out, avoiding unnecessary filtering steps and improving the overall performance. I've also seen multiple streets repeated 👀
>```
>NAV_DEBUG Queried names: [Paseo de Santa María de la Cabeza, Paseo de Santa María de la Cabeza, Autovía de Toledo, Autovía de Toledo]
>```

Could we include ☝️?

**Comment by danesfeder** [comment]
@Guardiola31337 yeah great call, the `WaynameFeatureFilter` now looks like:

```
  WaynameFeatureFilter(List<Feature> queriedFeatures, Location currentLocation, List<Point> currentStepPoints) {
    this.queriedFeatures = new ArrayList<>(new HashSet<>(queriedFeatures));
    this.currentPoint = Point.fromLngLat(currentLocation.getLongitude(), currentLocation.getLatitude());
    this.currentStepLineString = LineString.fromLngLats(currentStepPoints);
  }

```

Also, this is rebased and ready for another round 👀

**Pull request merged**